### PR TITLE
Store the typedef identifier in the symbol_type (cleaned up)

### PR DESCRIPTION
**Pull request by thk123**
Issue: https://github.com/diffblue/verification-engine-utils/issues/315

_This is an improved, tested and tidied version of this PR: https://github.com/diffblue/cbmc/pull/345, there is some useful discussion there._

The basic idea is to store the typedef in the type of each expression. Then when we convert a type, if it has an active typedef, we prefer that over the actual type. 

~This PR is dependent on: #353 for the tests to work as it uses the multi-line extension.~

~Real diff: https://github.com/diffblue/cbmc/pull/358/files/4437682559813e9efb9349416aaa764b8036b4a4..7d0a79335ae0d40f609a37a443b13ea6f1723082~

### CODE REVIEW

**Comment by thk123** [comment]
@peterschrammel I have replaced tabs with spaces in the new tests. Since they are all passing or marked as KNOWNBUG, I have added these folders to the root makefile so they will get run once that is turned back on.

**Comment by thk123** [comment]
One of the ANSI-C checks failed (`gcc_types_compatible_p1`) which seems to be a bug introduced by this PR so don't merge, I'll investigate what the problem is. 

There are some weird linting errors that I don't believe are on modified lines.

**Comment by thk123** [comment]
The problem was specifically in `c_typecheck_expr::gcc_types_compatible_p` where it is checking the types are exactly equal (which includes, for example, checking they have the same typedef'd name now). 

I have fixed this and added some extra tests. @peterschrammel could re-review these new changes?

File: src/ansi-c/c_typecheck_expr.cpp [hunk 1]
```diff
@@ -184,12 +186,13 @@ bool c_typecheck_baset::gcc_types_compatible_p(
...
   else
   {
-    if(type1==type2)
+    if(are_types_equal_ignoring_typedef(type1, type2))
     {
       // Need to distinguish e.g. long int from int or
       // long long int from long int.
       // The rules appear to match those of C++.
-
+      // Isn't this explictly handled by checking type1==type2 (since
```
Comment (peterschrammel) [inline src/ansi-c/c_typecheck_expr.cpp]: Do you mean "This isn't explicitly,,," (+typo)

Comment (thk123) [inline src/ansi-c/c_typecheck_expr.cpp]: No it was meant to be a question, meant to pull it out when I created the PR to ask the relevant person.

@kroening - is there a reason for [this line](https://github.com/diffblue/cbmc/pull/358/files/30fb4621075d69a61ec6974acd76674ada4b8b0a..23aa4262cfa1b26d19b17befcc728e68cbcfa5f6#diff-0984c6664412e8900073e1dd107c6658R196), it seems to me that it is implied by [this line](https://github.com/diffblue/cbmc/pull/358/files/30fb4621075d69a61ec6974acd76674ada4b8b0a..23aa4262cfa1b26d19b17befcc728e68cbcfa5f6#diff-0984c6664412e8900073e1dd107c6658L187)

**Comment by peterschrammel** [comment]
@thk123, please rebase.

**Comment by peterschrammel** [comment]
@thk123, seems that this needs rebasing again...

**Comment by thk123** [comment]
I have rebased but the regression tests now seem to be failing, going to see why that is now.

**Comment by tautschnig** [comment]
@thk123 see #467. Sorry for that one.

**Comment by thk123** [comment]
@tautschnig Ah cool, worrying that the master branch on Travis seemed to be passing the tests?

**Comment by peterschrammel** [comment]
This looks good to go. Merging #467 and #451 should make travis happy.

**Comment by thk123** [comment]
The regression tests are failing on this for some reason, need to investigate.

**Comment by thk123** [comment]
Regression tests are failing for a similar reason that led to this commit: https://github.com/diffblue/cbmc/pull/358/commits/1576ba02346fd605d3322d7da69b118274b69344. However, this has come up again (in multiplication of types this time). It seems to me that, like `ID_C_constant` the tag needs to go in the comments section of the `irept` so I will do that.

**Comment by thk123** [comment]
Fixed that issue and the original issue in a cleaner way (by just making the `ID_C_typedef` a comment on the structure rather than a named subtree.

@tautschnig: I could squash https://github.com/diffblue/cbmc/pull/358/commits/1576ba02346fd605d3322d7da69b118274b69344 with https://github.com/diffblue/cbmc/pull/358/commits/78dff22be803e73c6ce5a810aec1e733ebb4756f but I think this is maybe useful history, what do you think?

**Comment by tautschnig** [comment]
I do agree that this is a useful learning that may help understand the design decision, so, yes, don't squash those.

**Comment by thk123** [comment]
This is now failing as the regression make can't find folder `cbmc_typedef` will investigate.

**Comment by tautschnig** [comment]
@thk123 You don't have a Makefile in cbmc-typedef. (And I am starting to wonder whether the goto-instrument-typedef bits shouldn't just be moved to the goto-instrument folder.)

**Comment by thk123** [comment]
@tautschnig Ah yes, missed that in the checkin. I can combine into CBMC (or even goto-instrument since both support the show-symbol-table if you think that makes more sense?) I think my reason for splitting it out was for speed of running them independently and possibly to avoid the CBMC folder containing another 30 or so tests, if you prefer fewer high level folders I can combine.

**Comment by tautschnig** [comment]
I guess goto-instrument would make sense as a folder (technically it would be ansi-c, but then you can't use show-goto-functions on that one). Your consideration of execution time is very much appreciated, but we have so much cruft in the regression/ folder that I'd suggest to split up folders once all that has been cleaned up. Let's first have a few, clean, working regression folders, get rid of all the others, and then split them up again (with _all_ actively being part in regression suites).

**Comment by thk123** [comment]
This is failing on Mac - specifically the grep command used by the multi-line tests. This is because one of -Pzo flags are not supported. They are supposed to do:

- `-P` - enabled advanced Regex (I can't remember why this was needed, can probably be fixed to not use it)
- `-z` - treat whole string as one line (required, according to [this documentation](https://developer.apple.com/legacy/library/documentation/Darwin/Reference/ManPages/man1/grep.1.html) for Mac, this makes it behave like zgrep which appears to be for dealing with zipped structures
- `-o` - This only show matches, which is required since if it showed the line, it would be the whole line. This is supported by Macs grep.

I'm not sure how to fix this...

**Comment by tautschnig** [comment]
The best route forward might be to get rid of the external grep call, and use perl regex matching instead (whether using perl grep or not)?!

**Comment by thk123** [comment]
@tautschnig Yes, I feared as much, I just don't really know Perl so was hoping there was some way to avoid it... @jgwilson42 was talking about a more comprehensive solution to running the regression tests so pinging him to see if that is going to be any time soon.

**Comment by thk123** [comment]
@tautschnig #505 moves the Perl script to use the internal regex matching rather than passing off to grep and will therefore work consistently on Mac. Once that has been merged, I will rebase this PR to have that feature and adjust the tests to use the new multi line flag.

**Comment by thk123** [comment]
Rebased for Toyota release - still waiting on #505 so can fix the tests on Mac before can be merged. .

**Comment by thk123** [comment]
Rebased on 505 and passing - should now pass on Mac then this should be good to go.

**Comment by thk123** [comment]
@tautschnig Since #505 has been merged I think your block can be removed and this is good to go.

**Comment by tautschnig** [comment]
Indeed, approval set.

**Comment by thk123** [comment]
Fixed merge conflicts, assuming tests still pass this should be good to go.

**Comment by kroening** [comment]
There appear to be two different uses for ID_C_typedef in this PR: a) within the C frontend to denote the typedef type; b) outside of the C frotnend to annotate the symbol expressions.

Since extra IDs cost very little, I would suggest to use ID_typedef (as before) for purpose a), and ID_C_typedef for purpose b).

Then this seems ready to go!

**Comment by thk123** [comment]
@kroening In the original PR @tautschnig requested it be transformed into `ID_C_typedef` (https://github.com/diffblue/cbmc/pull/345#issuecomment-266249294). What does the `_C_` actually denote? I'm also not sure which uses are of type (a). Would it be then changes in `parser.y` and `parser.cpp`?

**Comment by kroening** [comment]
The _C_ is the equivalent of the # in the string representations; it means that these are 'comments' and not meant to carry semantic meaning. They are never used for IDs directly, only for keys.

**Comment by kroening** [comment]
I.e., you would revert the changes to parse.cpp, parser.y, c_storage_spec.cpp, ansi_c_convert_type.cpp

**Comment by thk123** [comment]
@kroening Thanks for clarifying - I believe this is done.

**Comment by thk123** [comment]
Currently @forejtv is looking at why this is failing on Windows.

**Comment by thk123** [comment]
@kroening All tests passing - this should be good to go.

File: .gitignore [hunk 2]
```diff
@@ -39,6 +39,7 @@ src/util/irep_ids.inc
 regression/**/tests.log
 regression/**/*.gb
 regression/**/*.smt2
+regression/**/*.gb
```
Comment (tautschnig) [inline .gitignore]: This duplicates an earlier line.

File: regression/Makefile [hunk 3]
```diff
@@ -1,10 +1,12 @@
 
 DIRS = ansi-c \
        cbmc \
+      cbmc-typedef \
```
Comment (tautschnig) [inline regression/Makefile]: The indent seems to be a bit off, unless that's just how github is printing this.

File: regression/Makefile [hunk 4]
```diff
@@ -1,10 +1,12 @@
...
        cpp \
        cbmc-java \
        goto-analyzer \
        goto-instrument \
+       goto-instrument-typedef \
```
Comment (tautschnig) [inline regression/Makefile]: Do we actually need to have these in new subfolders? Wouldn't the cbmc and goto-instrument folders do?

File: regression/Makefile (hunk 3 above)
Comment (thk123) [inline regression/Makefile]: This appears to be a GH rendering issue.

File: regression/Makefile (hunk 4 above)
Comment (thk123) [inline regression/Makefile]: For goto-instrument-typedef the chain.sh script does different things (specifically compiling with some flags to goto-cc). The cbmc ones look like they could be and I see I did agree to this before so will do.

File: src/ansi-c/ansi_c_parser.cpp [hunk 5]
```diff
@@ -215,7 +215,7 @@ Function: ansi_c_parsert::get_class
 
 ansi_c_id_classt ansi_c_parsert::get_class(const typet &type)
 {
-  if(type.id()==ID_typedef)
+  if(type.id()==ID_C_typedef)
```
Comment (kroening) [inline src/ansi-c/ansi_c_parser.cpp]: Would that not remain ID_typedef?

Comment (thk123) [inline src/ansi-c/ansi_c_parser.cpp]: Yes actually, I think so. This function doesn't seem to be called anywhere so the tests didn't cover it, perhaps it should just be deleted?

**Comment by thk123** [comment]
@kroening I have removed the `ID_C_typedef` change in ansi_c_parser.cpp as I believe you are right, but given that that code is not covered by any test or indeed called anywhere, it might be safer to delete it. I am happy to do this as a separate PR.

**Pull request merged**